R/params.R

```diff
@@ -0,0 +1,329 @@
+params_controls$Date <- function(param) { shiny::dateInput }
+## BUG: shiny does not support datetime selectors
+##     https://github.com/rstudio/shiny/issues/897
+##     we ask for string input for now.
+params_controls$POSIXct <- function(param) { shiny::textInput }
+params_controls$character <- function(param) { shiny::textInput }
+params_controls$data.frame <- function(param) { shiny::fileInput }
+params_get_control <- function(param) {
+  ## A value might have multiple classes. Try: class(Sys.time())
+  ## Try to find first class listed with a named control.
+  for (c in param$class) {
+    control <- params_controls[[c]]
+    if (!is.null(control)) {
+      return(control(param))
+    }
+  }
+  return(NULL)
+}
+
+params_configurable <- function(param) {
+  (!is.null(params_get_control(param))) && ( length(param$value) == 1 || "data.frame" %in% param$class )
+}
+
+#' Run a shiny application asking for parameter configuration for the given document.
+#'
+#' @param file Path to the R Markdown document with configurable parameters.
+#' @param input_lines Content of the R Markdown document. If \code{NULL}, the contents of \code{file} will be read.
+#' @param params A named list of optional parameter overrides used in place of the document defaults.
+#' @param shiny_args Additional arguments to \code{\link[shiny:runApp]{runApp}}.
+#'
+#' @return named list with overridden parameter names and value.
+#' 
+#' @export
+knit_params_ask <- function(file = "index.Rmd",
+                            input_lines = NULL,
+                            params = NULL,
+                            shiny_args = NULL,
+                            encoding = getOption("encoding")) {
+  if (packageVersion("knitr") < "1.10.17") {
+    stop("knitr >= 1.10.17 required to use rmarkdown::knit_params_ask")
+  }
+
+  if (is.null(input_lines)) {
+    input_lines <- read_lines_utf8(file, encoding)
+  }
+  
+  knit_params <- mark_utf8(knitr::knit_params(input_lines))
+  configurable <- Filter(params_configurable, knit_params)
+  unconfigurable <- Filter(Negate(params_configurable), knit_params)
+
+  ## Input validation on params (checks shared with render)
+  if (!is.null(params)) {
+    ## Must be a named list
+    if (!is.list(params) || (length(names(params)) != length(params))) {
+      stop("knit_params_ask params argument must be a named list")
+    }
+
+    ## Verify that all passed parameters are also in the yaml
+    invalid_params <- setdiff(names(params), names(knit_params))
+    if (length(invalid_params) > 0) {
+      stop("knit_params_ask params not declared in YAML: ",
+           paste(invalid_params, sep = ", "))
+    }
+  }
+  
+  ## Return the current value for a given parameter, which is either `params`
+  ## when overridden by our caller or `param$value` otherwise.
+  current_value <- function(param) {
+    if (!is.null(params)) {
+      override <- params[[param$name]]
+      if (!is.null(override)) {
+        return(override)
```

Comment:
This one (for `override`) is an explicit early return to prevent the final return of `param$value`. I used the `return` form for `param$value` to make both returned values explicit (it seemed to read better) but I can eliminate the return call for the default `param$value`. Is that your preference?